lib/msal-core/src/utils/RequestUtils.ts

```diff
@@ -140,7 +147,50 @@ export class RequestUtils {
      */
     static validateAndGenerateState(state: string): string {
         // append GUID to user set state  or set one for the user if null
-        return !StringUtils.isEmpty(state) ? CryptoUtils.createNewGuid() + "|" + state : CryptoUtils.createNewGuid();
+        return !StringUtils.isEmpty(state) ? RequestUtils.generateLibraryState() + "|" + state : RequestUtils.generateLibraryState();
+    }
+
+    /**
+     * Generates the state value used by the library.
+     *
+     * @returns Base64 encoded string representing the state
+     */
+    static generateLibraryState(): string {
+        const stateObject: StateObject = {
+            state: CryptoUtils.createNewGuid(),
+            ts: TimeUtils.now()
+        };
+
+        const stateString = JSON.stringify(stateObject);
+
+        return CryptoUtils.base64Encode(stateString);
+    }
+
+    /**
+     * Decodes the state value into a StateObject
+     *
+     * @param state State value returned in the request
+     * @returns Parsed values from the encoded state value
+     */
+    static parseLibraryState(state: string): StateObject {
+        const libraryState = state.split("|")[0];
+
+        if (CryptoUtils.isGuid(libraryState)) {
+            return {
+                state,
+                ts: TimeUtils.now()
+            };
+        }
```

Comment:
did we ever think of the potential use case for this? it feels like we should figure out if this is necessary or not, as it does affect how things work downstream

Reply:
Yes, a potential use case is a redirect flow: A user initiates a redirect action, and while they are away from the application, the application is updated to use the new format. When they return to the app, their state value would be a guid, and if this code isn't included, the request would fail. This would only happen once, but I think should be accounted for. This would require #1413 also having similar fallback logic to properly handle this use case.